### ISSUE TITLE
Doom Emacs adaptation of format-all

### ISSUE THREAD

**Issue by lassik** [lassik/emacs-format-all-the-code#19]
(Ping @hlissner from Doom Emacs)

So the [Doom Emacs](https://github.com/hlissner/doom-emacs/) project made an adaptation of this package in their `develop` branch. They have the right to do so, but unfortunately I don't agree with their decision at all.

Social issues:

* They didn't notify me about it.

* I heard about it by getting a bug report about a bug in the adaptation. So people are starting to get confused about different versions of format-all with different bugs.

* It seems they did it to address some issues that are not addressed in the main format-all package, but they didn't first talk to me about working together to fix those issues. That would fix them for all Emacs users, not just for Doom users.

* A strength of the Emacs package system (especially MELPA) is small modules that do one thing well. Then all the effort and wisdom on doing that thing can be gathered in that module and everyone benefits. As far as I can tell, Doom seems to be primarily an opinionated, pre-configured distribution like Spacemacs. Maintaining their own version of generic functionality like code formatting means that effort is duplicated fixing bugs and adding features in different packages. It would be healthier for the Emacs ecosystem if they only maintained a distribution, keeping the functionality in the existing small modules that can already be found in MELPA.

* In all respects, this seems to be fragmenting instead of uniting the Emacs community. I don't think that is wise, considering that Emacs probably already has less users and developers than Vim, VSCode and Atom. We should make the most of the time and energy of the contributors we have by centralizing one kind of functionality in one package.

Technical issues:

* I don't think it makes sense to provide a facility for users to define their own formatters. The ecosystem will be much healthier if everyone has the same formatter definitions, and improvements to those formatters are contributed back to the main package so everyone can benefit from that work. Format-all is open source, so users can add new formatters to the source code and make a pull request to include it in the main package.

* It is true that format-all doesn't support more than one formatter per language, or command line arguments for the formatters. [Unibeautify](https://github.com/Unibeautify/unibeautify) is a project to the right thing about this (providing a standard configuration file where formatters can be chosen and their settings be specified in one place that works with all editors and other tools). It's slower to develop but the end result is better, and also helps people who use other text editors besides Emacs. If people are really in a hurry to use command line arguments or multiple formatters, we could add a hook to format-all so people who really need them can use them with only a moderate amount of effort.

* Indenting regions of code (instead of a whole buffer) is a nice feature if it's done right. However,  at the moment there is practically no support from the external formatting programs we use. In my opinion, the right thing to do would be to add support to those programs. Otherwise the option is to do what the Doom adaptation does - insert and delete indentation without proper knowledge of the programming language it is dealing with. This is brittle and the right thing to do would be to talk to the people making the external formatters so they can implement it properly - not to make complicated hacks in front-ends like format-all.

* I decided against applying changes from formatters in RCS patch style because I looked at the code to do that in existing Emacs packages and it was very complex. I haven't had any issues with the current wipe-the-buffer-and-re-insert-the-whole-reformatted-code approach, but if you are very knowledgeable about this, let's work together on a solution involving the patches.

* They replaced format-all's macro-based definition facility (which was made to enable extremely compact formatter definitions, at the cost of one complex `defmacro`) with a new function-based one. I deliberately went in the other direction - the macro-based framework is a recent change and made the definitions simpler.

* Simplicity has been a key factor in all aspects of format-all. I looked at the existing code formatting packages for Emacs, and most of them were very complex and intricate for what they did, so I deliberately did everything possible in a simpler way. By contrast, the Doom adaptation seems to be going in the opposite direction, adding more features, not shying away from hacks.

I'd like to suggest the following to the people working on Doom Emacs:

* Please work together with format-all instead of making your own adaptation with lots of changes.
* Please work together to aggregate formatter definitions into format-all instead of each user and Emacs distribution having their own definitions.
* Please work together with Unibeautify like we do instead of doing an Emacs-only settings framework. I am already working with them and we'll be happy to have extra help to get Unibeautify done sooner.
* Please work together with the authors of external formatters to implement proper format-region functionality in the formatters themselves instead of doing indentation hacks in Emacs. I have reported bugs to some formatter authors and got a very good response from them. That fixes the problems for all the users of those formatters, not just Emacs.

In summary: let's do principled long-term technical work in collaboration as a wider community, instead of hacky duplicated short-term work divided into small groups who don't communicate.

Thanks for your consideration.

**Comment by mvdan** [comment]
My two cents - if forking is really the way to go here, the name of the software should be changed. Keeping the same name just confuses users and seems disrespectful to the original author, as now they will start getting bug reports they can't even fix.

**Comment by lassik** [comment]
lol, 3 thumbs up and 3 thumbs down :D Thanks to everyone who reacted, no matter what your stance.

Would the people who voted thumbs down please suggest alternative approaches? What do you think is wrong with the above and what would you do instead? I haven't yet got any verbal communication from anyone opposed to my suggestions or delivery, so I have no idea what you are thinking.

I get the impression that the intention is to change the name to something else but it's work in progress so they didn't get around to it yet. Also it's in the `develop` branch instead of `master`, but I'm told that `develop` is the recommended branch for users as well as developers.

**Comment by lassik** [comment]
Upon close reading, that code doesn't modify much of the core machinery of format-all, just extends it and the user interface, so the word choice "fork" was too strong on my part and I apologize for any harm it may have caused. I have changed it to the milder word "adaptation".

Anyway, my complaint is not about legal or attribution issues, but the fragmenting of functionality and interface across different packages and the use of hacks instead of sound long-term solutions. While the code is not a full-blown fork, the advice adds/changes so much functionality that it's effectively a different package while still bearing the same name. My wish is still that instead of combining bits and pieces of code for different distributions, we could try to do one package that caters to everyone's needs.

I could add a hook to the current format-all package so people can give command line arguments to formatters and chain together multiple formatters (#6). We could look into the RCS-patch-based buffer updates if you guys think that's a fundamentally better approach. And we can try to talk to the authors of external formatters about adding command line flags to format only parts of a buffer (#17).

The reason I'm reluctant to add customize variables for the main format-all package is that that would basically duplicate the effort of the Unibeautify project in an Emacs-only manner. They are trying to make one configuration file and framework that works the same way across all editors.

**Comment by CodyWilson** [comment]
> Would the people who voted thumbs down please suggest alternative approaches? What do you think is wrong with the above and what would you do instead? I haven't yet got any verbal communication from anyone opposed to my suggestions or delivery, so I have no idea what you are thinking.

I'll throw my $.02 in the ring.

If a set of functionality is desired/needed, and the package maintainer seems resistant to adding that functionality, why make a fuss when someone forks and does it in a "hacky" way?

Ultimately, if I took your repo, wrapped my own version of it that only I really use, stuck it in one of my repos, and went about my day. I would be guilty of "fragmenting of functionality and interface across different packages and the use of hacks instead of sound long-term solutions.".

**However**, I don't find that "fragmenting" to be a bad thing. I don't find quick "hacks" to accomplish something to *necessarily* be a bad thing.

Fragmenting is inevitable as people need different things from the same package, and expecting the central package to accommodate every possible "need", in my opinion, isn't going to generate "sound long-term solutions".

And when it comes to "hacks", virtually anything actually implementing a package to answers one of those "needs" is going to look like a hack. Because it's *very specific*.


One more thing.

> considering that Emacs probably already has less users and developers than Vim, VSCode and Atom.

There's no probably there. And I don't view that as much of a "problem" either. Not having a lot of users & developers doesn't make a product worse, it makes it niche.

**Comment by lassik** [comment]
Thank you for your thoughts, @CodyWilson.

> If a set of functionality is desired/needed, and the package maintainer seems resistant to adding that functionality, why make a fuss when someone forks and does it in a "hacky" way?

To the best of my knowledge they haven't asked me about adding or changing any features. If I seem resistant or difficult to approach, please tell me as that's not my intent.

> Ultimately, if I took your repo, wrapped my own version of it that only I really use, stuck it in one of my repos, and went about my day. I would be guilty of "fragmenting of functionality and interface across different packages and the use of hacks instead of sound long-term solutions.".

In my view that would depend on the popularity and scope of that repo. If it's just a quick hack for you and some friends, no problem. If it's a generic long-term package for thousands of users, I'd strongly favor wider collaboration.

I don't feel strongly enough to call anyone guilty :) I just feel bummed about a missed opportunity for more ambitious work that's almost within reach, as ambition is my fun and reward in life. That's all.

> **However**, I don't find that "fragmenting" to be a bad thing. I don't find quick "hacks" to accomplish something to _necessarily_ be a bad thing.

Me neither, in case there is long-term integrative work going on in parallel, but one of my pet peeves is the persistent culture in the software industry where hacks live on and the cleanup never comes. When hacks get the job done passably, it creates a dis-incentive to work on clean approaches as people start relying on the hacks and put more resources into developing and maintaining them.

I am guilty of this myself, as the maintenance of format-all delays the more important Unibeautify project by some amount. I try to keep this amount as small as I can and feed the knowledge gained from format-all into Unibeautify as it benefits a wider community.

> Fragmenting is inevitable as people need different things from the same package, and expecting the central package to accommodate every possible "need", in my opinion, isn't going to generate "sound long-term solutions".

For a really precise definition of *need* this is obviously true. However, I would imagine the user requirements for format-all and the Doom formatter to be in the same ballpark.

> And when it comes to "hacks", virtually anything actually implementing a package to answer one of those "needs" is going to look like a hack. Because it's _very specific_.

I'm not intimately familiar with this but I could imagine a situation where a package tailored to a specific formatter (such as the very complex *clang-format*) has very specific needs, whereas Doom seems more like a distribution, and formatting is just one of numerous functions it performs.

What I mean by a hack is when we work around a deficiency in some program or package instead of fixing the deficiency at its source in the original program or package. I've opened issues with several of the formatters supported by format-all and their maintainers have fixed (or are fixing) them, so we don't need to hack around those issues in format-all, improving both those formatters and format-all both in the short and long term. This response is very encouraging, I'm grateful to these maintainers for being so responsive and I'd love to collaborate more.

> Not having a lot of users & developers doesn't make a product worse, it makes it niche.

Sure, but more gets done if each problem is solved only once. Perhaps you imply that the social cost is too high; that may be right.

---

Perhaps the bigger picture here is that hackers often find it easier to do things themselves than talk to strangers and coordinate work across projects. This discussion has also made it clear that I should make a personal effort to be less disagreeable, as being too strict and having an attitude makes collaboration that much harder. I'll remind myself that being lenient is often more important than being right. I don't have any personal problem with anyone involved (on the contrary, I'm grateful to hear what people think instead of having to guess) as it's not that big of a deal and I have much bigger problems to worry about, but being perhaps too exacting about technical issues *is* a long-standing trait of mine that unfortunately causes fallout like this and I need to do something about that.

**Comment by lassik** [comment]
Closing this since there have been no new developments in a few months. For the record, I'm still open to collaboration with Doom and all other projects. I'm also trying to recruit new people to format-all and make it less of my personal project due to lack of time.
